toolchain/base/int_store.h

```diff
@@ -0,0 +1,186 @@
+// directly embedded in the `IntId` itself. Only larger integers are stored in
+// an array of `APInt` values and represented as an index in the ID.
+class IntStore {
+ public:
+  // Adds an integer value representable in a host `int64_t` to the store.
+  // Especially useful when the integer is computed without an `APInt` in the
+  // first place.
+  //
+  // This only accepts a signed `int64_t` and uses the mathematical signed
+  // integer value of it as the added integer value.
+  //
+  // Returns the ID corresponding to this integer value, storing an `APInt` if
+  // necessary to represent it.
+  auto Add(int64_t value) -> IntId {
+    // First try directly making this into an ID.
+    if (IntId id = IntId::TryMakeValue(value); id.is_valid()) [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return AddLarge(value);
+  }
+
+  // Stores a canonical copy of a signed value and returns its ID.
+  auto AddSigned(llvm::APInt value) -> IntId {
+    // First try directly making this into an ID.
+    if (IntId id = IntId::TryMakeSignedValue(value); id.is_valid()) [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return AddSignedLarge(std::move(value));
+  }
+
+  // Stores a canonical copy of an unsigned value and returns its ID.
+  auto AddUnsigned(llvm::APInt value) -> IntId {
+    // First try directly making this into an ID.
+    if (IntId id = IntId::TryMakeUnsignedValue(value); id.is_valid())
+        [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return AddUnsignedLarge(std::move(value));
+  }
+
+  // Returns the value for an ID.
+  //
+  // This will always be a signed `APInt` with a canonical bit width for the
+  // specific integer value in question.
+  auto Get(IntId id) const -> llvm::APInt {
+    if (id.is_value()) [[likely]] {
+      return llvm::APInt(MinAPWidth, id.AsValue(), /*isSigned=*/true);
+    }
+    return values_.Get(APIntId(id.AsIndex()));
+  }
+
+  // Returns the value for an ID adjusted to a specific bit width.
+  //
+  // Note that because we store canonical mathematical integers as signed
+  // integers, this always sign extends or truncates to the target width. The
+  // caller can then use that as a signed or unsigned integer as needed.
+  auto GetAtWidth(IntId id, int bit_width) const -> llvm::APInt {
+    llvm::APInt value = Get(id);
+    if (static_cast<int>(value.getBitWidth()) != bit_width) {
+      value = value.sextOrTrunc(bit_width);
+    }
+    return value;
+  }
+
+  // Returns the value for an ID adjusted to the bit width specified with
+  // another integer ID.
+  //
+  // This simply looks up the width integer ID, and then calls the above
+  // `GetAtWidth` overload using the value found for it. See that overload for
+  // more details.
+  auto GetAtWidth(IntId id, IntId bit_width_id) const -> llvm::APInt {
+    const llvm::APInt& bit_width = Get(bit_width_id);
```

Comment:
```suggestion
    const llvm::APInt bit_width = Get(bit_width_id);
```

Isn't the `&` incorrect here?

Reply:
FWIW, it is correct, but extremely misleading. Your suggestion is better.

The const reference binds and life-time extends the temporary returned by `Get`. Which is what we want, but a very subtle way of getting there.